#### 1. Imports and setup

In [2]:
import numpy as np
import pandas as pd
import plotly.express as px 

#### 2. Load dataset

In [3]:
initial_data = pd.read_csv('../data/raw/books_1.Best_Books_Ever.csv')

#### 3. High-Level Overview

In [4]:
initial_data.head()

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
0,2767052-the-hunger-games,The Hunger Games,The Hunger Games #1,Suzanne Collins,4.33,WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...,English,9780439023481,"['Young Adult', 'Fiction', 'Dystopia', 'Fantas...","['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...",...,NaN,['Locus Award Nominee for Best Young Adult Boo...,6376780,"['3444695', '1921313', '745221', '171994', '93...",96.0,"['District 12, Panem', 'Capitol, Panem', 'Pane...",https://i.gr-assets.com/images/S/compressed.ph...,2993816,30516,5.09
1,2.Harry_Potter_and_the_Order_of_the_Phoenix,Harry Potter and the Order of the Phoenix,Harry Potter #5,"J.K. Rowling, Mary GrandPré (Illustrator)",4.50,There is a door at the end of a silent corrido...,English,9780439358071,"['Fantasy', 'Young Adult', 'Fiction', 'Magic',...","['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...",...,06/21/03,['Bram Stoker Award for Works for Young Reader...,2507623,"['1593642', '637516', '222366', '39573', '14526']",98.0,['Hogwarts School of Witchcraft and Wizardry (...,https://i.gr-assets.com/images/S/compressed.ph...,2632233,26923,7.38
2,2657.To_Kill_a_Mockingbird,To Kill a Mockingbird,To Kill a Mockingbird,Harper Lee,4.28,The unforgettable novel of a childhood in a sl...,English,9999999999999,"['Classics', 'Fiction', 'Historical Fiction', ...","['Scout Finch', 'Atticus Finch', 'Jem Finch', ...",...,07/11/60,"['Pulitzer Prize for Fiction (1961)', 'Audie A...",4501075,"['2363896', '1333153', '573280', '149952', '80...",95.0,"['Maycomb, Alabama (United States)']",https://i.gr-assets.com/images/S/compressed.ph...,2269402,23328,NaN
3,1885.Pride_and_Prejudice,Pride and Prejudice,NaN,"Jane Austen, Anna Quindlen (Introduction)",4.26,Alternate cover edition of ISBN 9780679783268S...,English,9999999999999,"['Classics', 'Fiction', 'Romance', 'Historical...","['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...",...,01/28/13,[],2998241,"['1617567', '816659', '373311', '113934', '767...",94.0,"['United Kingdom', 'Derbyshire, England (Unite...",https://i.gr-assets.com/images/S/compressed.ph...,1983116,20452,NaN
4,41865.Twilight,Twilight,The Twilight Saga #1,Stephenie Meyer,3.60,About three things I was absolutely positive.\...,English,9780316015844,"['Young Adult', 'Fantasy', 'Romance', 'Vampire...","['Edward Cullen', 'Jacob Black', 'Laurent', 'R...",...,10/05/05,"['Georgia Peach Book Award (2007)', 'Buxtehude...",4964519,"['1751460', '1113682', '1008686', '542017', '5...",78.0,"['Forks, Washington (United States)', 'Phoenix...",https://i.gr-assets.com/images/S/compressed.ph...,1459448,14874,2.1


In [5]:
initial_data.describe()

,rating,numRatings,likedPercent,bbeScore,bbeVotes
count,52478.000000,5.247800e+04,51856.000000,5.247800e+04,52478.000000
mean,4.021878,1.787865e+04,92.231545,1.984023e+03,22.529003
std,0.367146,1.039448e+05,5.990689,3.515314e+04,369.158541
min,0.000000,0.000000e+00,0.000000,0.000000e+00,-4.000000
25%,3.820000,3.410000e+02,90.000000,8.400000e+01,1.000000
50%,4.030000,2.307000e+03,94.000000,9.700000e+01,1.000000
75%,4.230000,9.380500e+03,96.000000,1.870000e+02,2.000000
max,5.000000,7.048471e+06,100.000000,2.993816e+06,30516.000000


In [6]:
initial_data.shape

(52478, 25)

#### 4. Schema Inspection

In [7]:
initial_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 52478 entries, 0 to 52477
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   bookId            52478 non-null  str    
 1   title             52478 non-null  str    
 2   series            23470 non-null  str    
 3   author            52478 non-null  str    
 4   rating            52478 non-null  float64
 5   description       51140 non-null  str    
 6   language          48672 non-null  str    
 7   isbn              52478 non-null  str    
 8   genres            52478 non-null  str    
 9   characters        52478 non-null  str    
 10  bookFormat        51005 non-null  str    
 11  edition           4955 non-null   str    
 12  pages             50131 non-null  str    
 13  publisher         48782 non-null  str    
 14  publishDate       51598 non-null  str    
 15  firstPublishDate  31152 non-null  str    
 16  awards            52478 non-null  str    
 17  numR

The dataset contains 52478 rows and 25 columns. 20 columns are categorical and 5 are numeric. Genres, characters, awards, ratingsbystarts and setting are high-cardinality columns. Bookid is the ID column of the dataset. Some columns contain lists but have a Dtype of str.

#### 5. Missing Values Audit

In [8]:
missing = initial_data.isnull().mean().sort_values(ascending=False).mul(100).round(2)
missing

edition             90.56
series              55.28
firstPublishDate    40.64
price               27.37
language             7.25
publisher            7.04
pages                4.47
bookFormat           2.81
description          2.55
publishDate          1.68
likedPercent         1.19
coverImg             1.15
isbn                 0.00
ratingsByStars       0.00
bbeVotes             0.00
bbeScore             0.00
setting              0.00
awards               0.00
numRatings           0.00
genres               0.00
author               0.00
rating               0.00
title                0.00
characters           0.00
bookId               0.00
dtype: float64

Most books might not have a special edition. Missing series data is not meaningfull as not all books published are a part of a series. The missing price might need some investigation.

#### 6. Duplicate Analysis

In [9]:
duplicates = initial_data[initial_data.duplicated()]
duplicates.count()

bookId              50
title               50
series              25
author              50
rating              50
description         48
language            45
isbn                50
genres              50
characters          50
bookFormat          50
edition              6
pages               46
publisher           46
publishDate         49
firstPublishDate    27
awards              50
numRatings          50
ratingsByStars      50
likedPercent        49
setting             50
coverImg            50
bbeScore            50
bbeVotes            50
price               29
dtype: int64

In [10]:
no_duplicates = initial_data.drop_duplicates()
no_duplicates.shape

(52428, 25)

As only a small amount of rows were duplicates hence they were removed.

#### 7. Type and Parsing Problems

#### 8. Column-by-Column Audit

#### 9. Outlier Detection

#### 10. Data Quality Decisions

#### 11. Save Cleaned Intermediate Dataset